game-core/src/main/java/games/strategy/engine/data/properties/MapProperty.java

```diff
@@ -89,19 +87,19 @@ public boolean validate(final Object value) {
     if (value instanceof Map) {
       try {
         @SuppressWarnings("unchecked")
-        final Map<T, U> test = (Map<T, U>) value;
```

Comment:
I wonder if everything below could be replaced with [Collections#checkedMap](https://docs.oracle.com/javase/10/docs/api/java/util/Collections.html#checkedMap(java.util.Map,java.lang.Class,java.lang.Class)).
We don't really know what type to expect, but the code below seems to check if the objects are "compatible" with each other

Reply:
Interesting idea.  However, according to the Javadocs, it appears the type-checking only occurs when inserting a new value or modifying an existing value.  Looking at the implementation of `Collections#checkedMap()`, it doesn't appear that simply creating the wrapper does any type checking, which seems to be confirmed by this statement from the Javadocs:

> Assuming a map contains no incorrectly typed keys or values prior to the time a dynamically typesafe view is generated ...

Reply:
Interesting, I would have expected the constructor of the wrapper to check the map, but whatever.

Reply:
Second thought: Calling checkedMap and passing an empty map, to run putAll on it, or simply calling putAll with itself after the map has already been passed as a sort of verification.

Reply:
My static analyzer's flagging a potential NPE in this code, which I was going to follow-up on in another PR.  I'll investigate using `checkedMap()` further at that time.

Reply:
BTW, I don't think it will remove _all_ the code because we still have to determine what the value type actually is.  That's what the following code is doing (we can remove the key type determination and just use `String.class`):

```java
for (final Entry<T, U> entry : map.entrySet()) {
  if (entry.getValue() != null && entry.getKey() != null) {
    key = entry.getKey();
    val = entry.getValue();
    break;
  }
}
```